projects/extension/src/background/AppMediator.ts

```diff
@@ -146,34 +146,65 @@ export class AppMediator extends (EventEmitter as { new(): StateEmitter }) {
+    if (this.#pendingRequests.length > 0) {
+      this.#pendingRequests.forEach(req => this.#healthChecker?.sendJsonRpc(req));
+      this.#pendingRequests = [];
+    }
+  }
+
+  #handleError = (e: Error): void => {
+    this.#sendError(e.message);
+    this.#port.disconnect();
+    this.#manager.unregisterApp(this);
+  }
+
+  /** Handles the incoming message that contains Spec. */
+  #handleSpecMessage = (msg: MessageToManager, chainName: string): void => {
+    const chainSpec: string = relayChains.has(chainName) ?
+      (relayChains.get(chainName) || '') : msg.payload;
 
     const rpcCallback = (rpc: string) => {
-      const rpcResp = this.#healthChecker?.responsePassThrough(rpc);
+      const rpcResp: string | null | undefined = this.#healthChecker?.responsePassThrough(rpc);
       if (rpcResp)
-            this.#port.postMessage({ type: 'rpc', payload: rpcResp })
+        this.#port.postMessage({ type: 'rpc', payload: rpcResp })
     }
 
-    this.#manager.addChain(this.#chainName, chainSpec, rpcCallback)
-      .then(chain => {
-        this.#chain = chain;
-        // eslint-disable-next-line @typescript-eslint/unbound-method
-        this.#chain && this.#healthChecker?.setSendJsonRpc(this.#chain.sendJsonRpc);
-        this.#healthChecker?.start(this.#healthCheckCallback);
-        // process any RPC requests that came in while waiting for `addChain`
-        // to complete
-        if (this.#pendingRequests.length > 0) {
-          this.#pendingRequests.forEach(req => this.#healthChecker?.sendJsonRpc(req));
-          this.#pendingRequests = [];
-        }
-      })
-      .catch(e => {
-        this.#sendError((e as Error).message);
-        this.#port.disconnect();
-        this.#manager.unregisterApp(this);
-      });
+    // Means this is a parachain trying to connect
+    if (msg.parachainPayload) {
+      // Connect the main Chain first and on success the parachain with the chain
+      // that just got connected as the relayChain
+      const relayChainName: string = JSON.parse(msg.parachainPayload).relay_chain
+      const parachainSpec: string = msg.parachainPayload
+      const relaychainSpec: string | undefined = relayChains.get(relayChainName)
+
+      if (!relaychainSpec)
+        throw new Error('Relay chain spec was not found')
+
+      this.#manager.addChain(chainSpec).then(relayChain => {
+        this.#manager.addChain(parachainSpec, rpcCallback, relayChain).then(chain => {
+            this.#chain = chain;
+            this.#initHealthChecker();
+          }).catch(e => {
+            this.#handleError(e)
+          });
+        }).catch(e => {
+          this.#handleError(e)
+        });
+    } else {
+      // Connect the main Chain only
+      this.#manager.addChain(chainSpec, rpcCallback)
+        .then(chain => {
+          this.#chain = chain;
+          this.#initHealthChecker();
+        }).catch(e => {
+          this.#handleError(e)
+        });
+    }
```

Comment:
A bit of a nit-pick, but IMO this part could be simplified a bit:

```suggestion
    let chainPromise: Promise<SmoldotChain>
    
    // Means this is a parachain trying to connect
    if (msg.parachainPayload) {
      // Connect the main Chain first and on success the parachain with the chain
      // that just got connected as the relayChain
      const relayChainName: string = JSON.parse(msg.parachainPayload).relay_chain
      const parachainSpec: string = msg.parachainPayload
      const relaychainSpec: string | undefined = relayChains.get(relayChainName)

      if (!relaychainSpec)
        throw new Error('Relay chain spec was not found')

      chainPromise =  this.#manager.addChain(chainSpec).then(relayChain =>
        this.#manager.addChain(parachainSpec, rpcCallback, relayChain)
      )
    } else {
      // Connect the main Chain only
      chainPromise = this.#manager.addChain(chainSpec, rpcCallback)
    }
    
    chainPromise.then(chain => {
      this.#chain = chain;
      this.#initHealthChecker();
    }).catch(e => {
      this.#handleError(e)
    });
```